Source/WebCore/rendering/RenderBlockFlow.cpp

```diff
@@ -2998,6 +2998,25 @@ std::optional<LayoutUnit> RenderBlockFlow::firstLineBaseline() const
     return LayoutUnit { firstRootBox()->logicalTop() + firstLineStyle().metricsOfPrimaryFont().ascent(firstRootBox()->baselineType()) };
 }
 
+std::optional<LayoutUnit> RenderBlockFlow::lastLineBaseline() const
+{
+    if (isWritingModeRoot() && !isRubyRun() && !isGridItem())
+        return std::nullopt;
+
+    if (shouldApplyLayoutContainment())
+        return std::nullopt;
+
+    if (!childrenInline())
+        return RenderBlock::lastLineBaseline();
+
+    if (!hasLines())
+        return std::nullopt;
+
+    const LayoutIntegration::LineLayout* lineLayout = modernLineLayout();
+    ASSERT(lineLayout); 
+    return LayoutUnit { floorToInt(lineLayout->lastLineLogicalBaseline()) };
```

Comment:
Why does this use lastLineLogicalBaseline and  RenderBlockFlow::firstLineBaseline uses firstLinePhysicalBaseline? What about the last few lines of RenderBlockFlow::firstLineBaseline, why is that not needed here?

Reply:
We need a LineLayout::lastLinePhysicalBaseline() here and it should also be moved behind the LineBox iterator interface.
1. you can't assume all the content goes through the modern line layout codepath
2. It would also eliminate the need for duplicating the vertical flip (what Rob refers to) since that would be moved behind the iterator as well.
Not sure why we haven't done that yet.

Reply:
I implemented LineLayout::lastLinePhysicalBaseline and added some extra logic for the case where the code goes through legacy layout. After talking with Alan, it seems like this should be ok for now, but eventually we will want to do this refactoring. Things will also probably change as we move more away from legacy line layout.